src/components/leaderboards/Create.vue

```diff
@@ -0,0 +1,144 @@
+<script setup lang="ts">
+import { ref } from 'vue'
+import { onBeforeRouteLeave, useRouter } from 'vue-router'
+import { useApi } from '../../composables/useApi'
+import { useAuth } from '../../composables/useAuth'
+import { useSessionToken } from '../../composables/useSessionToken'
+import { Leaderboards } from '../../lib/api/Leaderboards'
+import { CreateLeaderboardRequest } from '../../lib/api/data-contracts'
+
+const token = useSessionToken()
+const createError = ref('')
+const router = useRouter()
+
+const leaderboards = new Leaderboards({
+	baseUrl: import.meta.env.VITE_BACKEND_URL
+})
+
+const warnBeforeLeave = ref(true)
+
+const createRequest = ref<CreateLeaderboardRequest>({
+	name: '',
+	slug: '',
+	info: '',
+})
+
+onBeforeRouteLeave(() => {
+	if (
+		warnBeforeLeave.value &&
+		(createRequest.value.name ||
+		 createRequest.value.slug ||
+		 createRequest.value.info)
+	) {
+		if (!window.confirm('Do you want to leave? You have unsaved changes.')) {
+			return false
+		}
+	}
+})
+
+async function submit() {
+	useApi(
+		() =>
+			leaderboards.createLeaderboard(
+				createRequest.value,
+				useAuth(token.value)
+			),
+		() => {
+			warnBeforeLeave.value = false
+			router.push({ name: 'leaderboardsList' })
+		},
+		(error) => {
+			createError.value = 'Failed to create: ' + error.status.toString(10)
+		}
+	)
+}
+</script>
+
+<template>
+	<div class="container">
+		<h1>Create Leaderboard</h1>
+		<RouterLink
+			class="back-link"
+			:to="{ name: 'leaderboardsList' }"
+			title="Changes will not be saved."
+		>
+			&lt; Back
+		</RouterLink>
+
+		<p v-if="createError" class="error-text">{{ createError }}</p>
+
+		<form @submit.prevent="submit">
+			<p class="info">Fields marked * are mandatory.</p>
+			<table>
+				<tbody>
+					<tr>
+						<th>
+							<label for="name">Name*:</label>
+						</th>
+						<td>
+							<input required v-model="createRequest.name" id="name" />
+						</td>
+					</tr>
+					<tr>
+						<th>
+							<label for="slug">Slug*:</label>
+						</th>
+						<td>
+							<input required pattern="^[a-zA-Z0-9\-_]{2,80}$" v-model="createRequest.slug" id="slug" />
```

Comment:
This is super obscure html knowledge but the caret and dollar sign aren't needed here. Also, double-check if the backslash needs to be double-escaped or not.

Reply:
The backslash is for the dash, right? I've just tested it (don't know why I didn't think to) and we shouldn't need to escape the dash at all.

Reply:
Hyphens indicate ranges so some regex parsers require escaping them.

Reply:
Yeah I understand that but we don't need to in JS. It interprets it correctly.